extensions/wikia/ContributionAppreciation/ContributionAppreciationController.class.php

```diff
@@ -49,9 +93,55 @@ public static function onPageHistoryBeforeList() {
 		return true;
 	}
 
+	public static function onBeforePageDisplay( \OutputPage $out, \Skin $skin ) {
+		if ( self::shouldDisplayApprectiation() ) {
+			Wikia::addAssetsToOutput( 'contribution_appreciation_user_js' );
+		}
+
+		return true;
+	}
+
 	private static function shouldDisplayApprectiation() {
 		global $wgUser, $wgLang;
 
 		return $wgUser->isLoggedIn() && $wgLang->getCode() === 'en';
+
+	}
+
+	private function prepareAppreciations( $upvotes ) {
+		$appreciatons = [];
+
+		foreach( $upvotes as $upvote ) {
+			$wikiId = $upvote['revision']['wikiId'];
+			$title = GlobalTitle::newFromId( $upvote['revision']['pageId'], $wikiId );
+
+			if ( !$title instanceof Title ) {
+				continue;
+			}
+
+			$diffLink = $this->getDiffLink( $title, $upvote['revision']['revisionId'] );
+
+			$userLinks = [];
+			foreach ( $upvote['upvotes'] as $userUpvote ) {
+				$userLinks[] = $this->getUserLink( $userUpvote['from_user'], $wikiId );
+			}
+
+			if ( !empty( $userLinks ) ) {
+				$appreciatons[] = wfMessage( 'appreciation-user', count( $userLinks ) )->rawParams(
+					implode( ', ', $userLinks ),
+					$diffLink
+				)->escaped();
+			}
+		}
+
+		return $appreciatons;
+	}
+
+	private function getDiffLink( Title $title, $revisionId ) {
+		return Html::element( 'a', [
+			'href' => $title->getFullURL( [ 'diff' => $revisionId, 'oldid' => 'prev' ] ),
+			'data-tracking' => 'notification-diff-link',
+			'target' => '_blank'
```

Comment:
I think yes. If user gets more than one appreciation, click this link and goes to diff page without opening it in new tab, then there is no way right now to come back and see other appreciations.